src/promptflow-core/promptflow/_utils/async_utils.py

```diff
@@ -22,6 +24,59 @@ def _has_running_loop() -> bool:
         return False
 
 
+class _AsyncTaskSigIntHandler:
+    """The handler to cancel the current task if SIGINT is received.
+    This is only for python<3.11 where the default cancelling behavior is not supported.
+    The code is similar to the python>=3.11 builtin implementation.
+    https://github.com/python/cpython/blob/46c808172fd3148e3397234b23674bf70734fb55/Lib/asyncio/runners.py#L150
+    """
+
+    def __init__(self, task: asyncio.Task, loop: asyncio.AbstractEventLoop):
+        self._loop = loop
+        self._task = task
+        self._interrupt_count = 0
+
+    def on_sigint(self, signum, frame):
+        self._interrupt_count += 1
+        if self._interrupt_count == 1 and not self._task.done():
+            self._task.cancel()
+            # This call_soon_threadsafe would schedule the call as soon as possible,
+            # it would force the event loop to wake up then handle the cancellation request.
+            # This is to avoid the loop blocking with long timeout.
+            self._loop.call_soon_threadsafe(lambda: None)
+            return
+        raise KeyboardInterrupt()
+
+
+async def _invoke_async_with_sigint_handler(async_func, *args, **kwargs):
+    """In python>=3.11, the asyncio.run in default cancel the running tasks when sigint is hit.
+    This is very useful when we call async functions.
+    So we have a similar implementation here to make sure python>3.11 also have such feature.
```

Comment:
"make sure python>3.11 also have such feature."
should this be "python<3.11" ?

Reply:
Good catch, updated.